compaction/compaction_manager.cc

```diff
@@ -548,12 +554,6 @@ requires (compaction_manager& cm, Args&&... args) {
 future<compaction_manager::compaction_stats_opt> compaction_manager::perform_compaction(throw_if_stopping do_throw_if_stopping, std::optional<tasks::task_info> parent_info, Args&&... args) {
     auto task_executor = seastar::make_shared<TaskExecutor>(*this, std::forward<Args>(args)...);
     gate::holder gate_holder = task_executor->_compaction_state.gate.hold();
-    _tasks.push_back(task_executor);
-    auto unregister_task = defer([this, task_executor] {
-        _tasks.remove(task_executor);
-        task_executor->switch_state(compaction_task_executor::state::none);
-        task_executor->release_resources();
-    });
 
     if (parent_info) {
         auto task = co_await get_task_manager_module().make_task(task_executor, parent_info.value());
```

Comment:
if compaction manager is asked to remove a table (which means stopping all tasks associated with that table), wouldn't it miss a task that is currently sleeping in task manager's make_task (which can preempt)? Then it's undefined behavior what happens thereafter as table could be dropped by then.

You said that motivation for patch is "Registering them in compaction manager earlier can lead to undesired behaviour, e.g. appearing as they are complete". Cannot we special case tasks that are in pending state and show them as they haven't started yet?

Reply:
My motivation for the patch is quite vague. What I really meant was that there may be situation where `stop_tasks` won't wait for the compaction because compaction_done() will be resolved immediately even though it shouldn't.

Getting into details:
We have 3 main operations here:
1. calling make_task to register it in task manager (preempts)
2. insert task into compaction_manager::_tasks
3. compaction_task_executor::_compaction_done = do_run()

Let's try to order them.

First let's see that there can be no preemption betweem 2 and 3:
- 2 -> preemption (with compaction_manager::stop_ongoing_compactions call) -> 3 -
	2 
	compaction_task_executor::_compaction_done has a default value which is a ready future
	compaction_manager::stop_ongoing_compactions stops tasks and waits for them to be complete (`co_await task->compaction_done();` in `stop_tasks`)
	3 - and task runs even though it was supposed to stop
^ this is a situation which I was trying to avoid with the patch
- 3 -> preemption (with compaction_manager::stop_ongoing_compactions call) -> 2
	3
	compaction_manager::stop_ongoing_compactions does not know about the task so it's not stopped even though it is supposed to.
	2

So, let's consider them as a group [23] (order of the two does not matter) and try to find a good way for 1.

A) [23] -> 1
Compaction may be running while it isn't registered in task manager yet. Quite not hurtful. Quite not what we aim for.

1 -> [23]
Here's the problem you have mentioned. Actually I think it won't be UB but a really long table removal since in compaction_manager::remove we wait until 
`co_await seastar::when_all_succeed(stop_ongoing_compactions("table removal", &t), c_state.gate.close()).discard_result();` and the gate won't close until the unlucky compaction finishes and leaves the gate. Nevertheless, it's not acceptable.

To summarize: the only remaining option here is A), however it's rather not satisfying.

I think this leads to conclusion that composition, and not inheritance, of compaction_task_executor is the way and it shouldn't be postponed anymore. (@bhalevy what do you think?)

Reply:
I see. Maybe we can use compaction_task_executor::stopping() to decide if we want to proceed with compaction. Alternatively, we play with promise and future, where default becomes non ready future and resolved once task either fails or succeed, but that may add unneeded complexity. And that means relying on compaction starting to actually stop the task, and that's not needed.

Reply:
That's right, thanks, I don't know why abandoned other ways to fix 213.